### PR TITLE
Adding QR decomposition

### PR DESCRIPTION
**Pull request by anagontijo**
## Description

Please include a summary of the change and which issue/s is/are fixed.
Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes: #343

Changes proposed:
- Norm calculator for vectors using numeric method
- Reflection calculator for vectors using householder reflector
- QR decomposition calculators

## Type of change

Select relevant options.

[ ] Bug fix (non-breaking change which fixes an issue)
[X] New feature (non-breaking change which adds functionality)
[ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
[ ] Documentation update

Are all split configurations tested and accounted for?
[X] yes [ ] no
Does this change require a documentation update outside of the changes proposed?
[X] yes [ ] no
Does this change modify the behaviour of other functions?
[ ] yes [X] no
Are there code practices which require justification?
[ ] yes [X] no

### CODE REVIEW

**Comment by coquelin77** [comment]
Unit test are required for all PRs. Please add these to test_linalg.py file in the tests folder. Also, code must follow pep8 formatting. See the contributing.md file in the home directory. 

An algorithmic note: when this method was implemented previously it showed very poor scaling results. are you seeing this as well?

File: heat/core/linalg.py [hunk 1]
```diff
@@ -877,3 +877,87 @@ def triu(m, k=0):
         Upper triangle of the input tensor.
     """
     return __tri_op(m, k, torch.triu)
+
+def norm(w,norm_order=2):
+    """
+        Calculates the norm of a given vector
+    """
+    beta = abs(max(w[0], key=abs))
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: These should probably be methods of heat, i.e. ht.abs()/DNDarray.abs() and ht.max()/DNDarray.max().

File: heat/core/linalg.py [hunk 2]
```diff
@@ -877,3 +877,87 @@ def triu(m, k=0):
         Upper triangle of the input tensor.
     """
     return __tri_op(m, k, torch.triu)
+
+def norm(w,norm_order=2):
+    """
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: Does not conform to doc strings of project

File: heat/core/linalg.py [hunk 3]
```diff
@@ -877,3 +877,87 @@ def triu(m, k=0):
...
+    return x_norm
+
+def householder_reflect(w):
+    """
+        Calculates the reflection of a vector w using the householder reflector
+    """
+    tau = norm(w.T)
+
+    if tau == 0 or w.shape[0] == 1:
+        return 0, np.zeros((w.shape)), 0, np.identity((w.shape[0]))
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: There should not be any numpy code here. Instead it should (preferably) be ht.zeros and ht.eye calls or at least torch.zeros and torch.eye to allow GPU computations

File: heat/core/linalg.py [hunk 4]
```diff
@@ -877,3 +877,87 @@ def triu(m, k=0):
...
+
+    tau_vec = np.zeros((w.shape))
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: See numpy calls issue above

File: heat/core/linalg.py [hunk 5]
```diff
@@ -877,3 +877,87 @@ def triu(m, k=0):
...
+        beta = 1
+    v = [element/beta for element in w[0]]
+
+    v_norm = (sum(element**norm_order for element in v))**(1/norm_order)
+    x_norm = v_norm * beta
+
+    return x_norm
+
+def householder_reflect(w):
+    """
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: Project doc strings

File: heat/core/linalg.py [hunk 6]
```diff
@@ -877,3 +877,87 @@ def triu(m, k=0):
...
+def norm(w,norm_order=2):
+    """
+        Calculates the norm of a given vector
+    """
+    beta = abs(max(w[0], key=abs))
+    if beta == 0:
+        beta = 1
+    v = [element/beta for element in w[0]]
+
+    v_norm = (sum(element**norm_order for element in v))**(1/norm_order)
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: Python internal sum call, should be an ht call

File: heat/core/linalg.py [hunk 7]
```diff
@@ -877,3 +877,87 @@ def triu(m, k=0):
...
+    v = w - tau_vec
+
+    gamma = 2 / norm(v.T)**2
+
+    matrix = np.identity(w.shape[0]) - gamma*(v.dot(v.T))
+
+    return gamma, v.T, tau, matrix
+
+def qr_decompose(A):
+    """
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: Project doc string format

File: heat/core/linalg.py [hunk 8]
```diff
@@ -877,3 +877,87 @@ def triu(m, k=0):
...
+        Decomposes a matrix A in the matrixes Q and R such that Q*R = A and R is a right superior triangular matrix
+    """
+    # Getting the matrix shape
+    n, m = A.shape
+    # Initializing Q as an identity matrix
+    q = np.identity(n)
```
Comment (Markus-Goetz) [inline heat/core/linalg.py]: See numpy calls issue above

**Comment by Markus-Goetz** [comment]
Stale

**Pull request closed without merging**